### DISCUSSION TITLE
Help implementing Time Filters

### DISCUSSION THREAD

**Discussion by mbellass** [Hugovdberg/PIconnect#609] [Q&A]
Hello!

I am looking to get the amount of time a piece of equipment is in the "RUNNING" state. I tried to perform this using filtered summaries, but it appears the filtered summaries will not work for non-numeric tags.

Currently I use "Time Filtered" function in excel with an expression '%tag%' = "RUNNING", to get the running hours on the equipment between two dates. I would like to achieve this with PIconnect.

I have done some research and found that the IPICalculation (part of the PI SDK) looks like it would achieve what I am looking for, but I am not familiar with SDKs.

Has anyone else implemented "Time Filtered" expressions with non-numeric expressions in PIconnect?

Reference:
https://pisquare.osisoft.com/s/question/0D51I00004UHgD1SAL/pi-sdk-filtered-summaries-total

Thank you!

**Comment by Hugovdberg** [comment]
`IPICalculation` is an old interface from the PI SDK (predecessor to the current AF SDK), as documented on [this page](https://docs.osisoft.com/bundle/af-sdk/page/html/pi-sdk-equivalents.htm). That interface was replaced by the [`AFCalculation`](https://docs.osisoft.com/bundle/af-sdk/page/html/T_OSIsoft_AF_Data_AFCalculation.htm) class.

I did try to use it, but can't get it to work yet. Unfortunately I don't have much time before my holidays so I probably won't get around to this for the next few weeks. Perhaps the following snippet is helpful to you to figure out how to get this working. 

```python
import PIconnect as PI
from PIconnect.PIConsts import TimestampCalculation

percent_true = PI.AF.Data.AFCalculation.PercentTrue
parse_interval = PI.AF.Time.AFTimeSpan.Parse

with PI.PIAFDatabase() as database:
    elem = database.descendant("Path\\To\\Element")
    point = elem.attributes["AttributeName"]
    expression = "'{tag}' = \"RUNNING\""
    time_range = PI.time.to_af_time_range("-2d", "T") # Calculate over the previous 2 days
    calculation_interval = parse_interval("1h") # Return summaries for each hour
    min_sample_interval = parse_interval("5m") # Determine status at least every 5 minutes
    time_type = int(TimestampCalculation.EARLIEST_TIME) # Return value at the start of each calculation interval
    return_values = PI.AF.Asset.AFValues()
    print(point, expression.format(tag=point.name))
    print(percent_true(point.attribute, expression.format(tag=point.name), time_range, calculation_interval, min_sample_interval, time_type, return_values))
```

This raises the following error:

```
---------------------------------------------------------------------------
InvalidOperationException                 Traceback (most recent call last)
~\AppData\Local\Temp/ipykernel_17356/1046204202.py in <module>
     15     return_values = PI.AF.Asset.AFValues()
     16     print(point, expression.format(tag=point.name))
---> 17     print(percent_true(point.attribute, expression.format(tag=point.name), time_range, calculation_interval, min_sample_interval, time_type, return_values))

InvalidOperationException: Unrecognized variable 'AttributeName' in expression.
   bij OSIsoft.AF.Data.AFCalculation.ConvertVariable(Object sourceTarget, Object& destinationTarget, String variable, Boolean variableRequired, Boolean useDataReferenceSyntax, Boolean& autoDeterminePIServer)
   bij OSIsoft.AF.Data.AFCalculation.ConvertExpressionTargetInternal(Object sourceTarget, Object& destinationTarget, String expression, Boolean useDataReferenceSyntax, Boolean autoDeterminePIServer)
   bij OSIsoft.AF.Data.AFCalculation.ConvertExpressionToPIServer(Object sourceTarget, PIServer& server, String expression)
   bij OSIsoft.AF.Data.AFCalculation.PercentTrue(Object target, String expression, AFTimeRange timeRange, AFTimeSpan calculationInterval, AFTimeSpan minimumSampleInterval, AFTimestampCalculation timeType, AFValues& percentGood)
```

**Reply by mbellass** [reply]
Thank you Hugovdberg. This helped tremendously!

I was able to get your code above working by adding the server address to the expression.

Let me know if you think others are interested and I can share how I expanded on this.

Thank you!

**Reply by Hugovdberg** [reply]
It is always nice to have the package reflect more of the functionality of the underlying SDK, so if you would create a pull request that adds the function then I'm more than willing to merge it! Let me know if you need help to implement it!
